stan/math/laplace/laplace_marginal.hpp

```diff
@@ -0,0 +1,275 @@
+#ifndef STAN_MATH_LAPLACE_LAPLACE_MARGINAL_HPP
+#define STAN_MATH_LAPLACE_LAPLACE_MARGINAL_HPP
+
+#include <stan/math/prim/mat/fun/Eigen.hpp>
+#include <stan/math/prim/mat/fun/quad_form_diag.hpp>
+#include <stan/math/prim/mat/fun/diag_pre_multiply.hpp>
+#include <stan/math/prim/mat/fun/diag_post_multiply.hpp>
+#include <stan/math/prim/mat/fun/cholesky_decompose.hpp>
+#include <stan/math/prim/mat/fun/sqrt.hpp>
+#include <stan/math/rev/mat/fun/cholesky_decompose.hpp>
+#include <stan/math/laplace/laplace_likelihood.hpp>
+
+
+// Reference for calculations of marginal and its gradients:
+// Rasmussen and Williams,
+// "Gaussian Processes for Machine Learning", first edition,
+// Algorithms 3.1 and 5.1.
+// The MIT Press, 2006.
+// Note: where I didn't conflict with my own notation, I used their notation.
+
+namespace stan {
+namespace math {
+  /**
+   * Function to compute the log marginal density.
+   * Variables needed for the gradient are stored by reference.
+   *
+   * theta_0: initial guess for latent gaussian variable.
+   * phi: global parameter (input in covariance function)
+   * x: fixed data (input in covariance function)
+   * D: structure to compute the conditional log likelihood and its
+   *    derivatives.
+   * K: a functor to compute the covariance function specified by the user.
+   * theta: mode of the target.
+   * W_root: squared root of the negative diagnonal Hessian.
+   * L: Cholesky decomposition of B, the stabilized covariance inverse.
+   * a: component in the Newton step.
+   * tolerance: function tolerance defined in terms of objective function.
+   * max_num_steps: number of Newton iterations after which the algorithm
+   *                stops.
+   */
+  template <typename D, typename K>
+  double
+  laplace_marginal_density (const Eigen::VectorXd& theta_0,
+                            const Eigen::VectorXd& phi,
+                            const std::vector<Eigen::VectorXd>& x,
+                            const D& diff_likelihood,
+                            const K& covariance_function,
+                            Eigen::VectorXd& theta,
+                            Eigen::VectorXd& W_root,
+                            Eigen::MatrixXd& L,
+                            Eigen::VectorXd& a,
+                            Eigen::VectorXd& l_grad,
+                            double tolerance = 1e-6,
+                            long int max_num_steps = 100) {
+    using Eigen::MatrixXd;
+    using Eigen::VectorXd;
+
+    int group_size = theta_0.size();
+    MatrixXd covariance = covariance_function(phi, x, group_size);
+    theta = theta_0;
```

Comment:
`laplace_marginal` is not exposed to the user. Changing and saving the arguments allows us to reuse some of the computation inside the chain method.

Reply:
> laplace_marginal is not exposed to the user. 

It's not exposed to the user but it is exposed to other stan devs (who once you write this will also be users) :)

> Changing and saving the arguments allows us to reuse some of the computation inside the chain method.

I need to look into how much savings we get from this